lib/argon2.js

```diff
@@ -58,63 +54,135 @@
      *      .then(h => console.log(h.hash, h.hashHex, h.encoded))
      *      .catch(e => console.error(e.message, e.code))
      */
+
     function argon2Hash(params) {
-        if (!scriptLoadedPromise) {
-            scriptLoadedPromise = loadScript(params.asmFileUrl);
+        if (!global.WebAssembly) {
+            return new Promise(function (resolve, reject) {
+                loadScript(params.asmFileUrl)
+                    .then(function () {
+                        resolve(calcHash(params));
+                    })
+                    .catch(function (error) {
+                        reject("Error loading argon2-asm.min.js: " + error + ", " + params.cacheKey);
+                    });
+            })
         }
-        return scriptLoadedPromise.then(function () {
-            var tCost = params.time || 1;
-            var mCost = params.mem || 1024;
-            var parallelism = params.parallelism || 1;
-            var pwd = allocateArray(params.pass);
-            var pwdlen = params.pass.length;
-            var salt = allocateArray(params.salt);
-            var saltlen = params.salt.length;
-            var hash = Module.allocate(new Array(params.hashLen || 24), 'i8', Module.ALLOC_NORMAL);
-            var hashlen = params.hashLen || 24;
-            var encoded = Module.allocate(new Array(512), 'i8', Module.ALLOC_NORMAL);
-            var encodedlen = 512;
-            var argon2Type = params.type || ArgonType.Argon2d;
-            var version = 0x13;
-            var err;
-            try {
-                var res = Module._argon2_hash(tCost, mCost, parallelism, pwd, pwdlen, salt, saltlen,
-                    hash, hashlen, encoded, encodedlen, argon2Type, version);
-            } catch (e) {
-                err = e;
+
+        //WebAssembly is supported by the browser
+
+        return new Promise(function (resolve, reject) {
+            if (global.Module && !global.Module._argon2_hash) {
+                setTimeout(function () { argon2Hash(params).then(resolve).catch(reject) }, 10)
```

Comment:
Could you add a comment above somewhere here to explain why you are doing this? It's not obvious.